### PR TITLE
Fix publish issue with small RDP constant values

### PR DESCRIPTION
**Pull request by IshanMi**
## Description
This solves the issue raised here: #6949 

The problem was that when we had rdp_constants less than 1, the `convert_constants_to_indices` function would return a negative index. When calling `jnp.take` on the cache with negative indices, the epsilon spend would be gigantic, and this would prevent publish from happening.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: packages/syft/src/syft/core/adp/vectorized_publish.py [hunk 1]
```diff
@@ -156,7 +156,15 @@ def publish(
 
         # rdp_constant = all terms in Theorem. 2.7 or 2.8 of https://arxiv.org/abs/2008.11193 EXCEPT alpha
         rdp_constants = compute_rdp_constant(rdp_params, private=private)
-        print("Rdp constants", rdp_constants)
+        if any(rdp_constants < 0):
+            raise Exception(
+                "Negative budget spend not allowed in PySyft for safety reasons."
+                "Please contact the OpenMined support team for help."
```
Comment (iamtrask) [inline packages/syft/src/syft/core/adp/vectorized_publish.py:162]: Let's indicate where/how they'd contact us. (slack, email, github issue, or otherwise)

**Review by iamtrask — DISMISSED** [review]

File: packages/syft/src/syft/core/adp/vectorized_publish.py [hunk 2]
```diff
@@ -156,7 +156,15 @@ def publish(
...
+            )
+        if any(np.isnan(rdp_constants)) or any(np.isinf(rdp_constants)):
+            raise Exception(
+                "Invalid privacy budget spend. Please contact the OpenMined support team for help."
```
Comment (teo-milea) [inline packages/syft/src/syft/core/adp/vectorized_publish.py:166]: This is alright for the DO, but on the DS side, the publishing returns no error, and .get() returns the object not found error. We need to find a way to signal unsolved publishing to the DS, without it being a privacy leak.

Comment (teo-milea) [reply packages/syft/src/syft/core/adp/vectorized_publish.py:166]: This is something that we will do in the next release, will close this for now.

File: packages/syft/src/syft/core/adp/data_subject_ledger.py [hunk 3]
```diff
@@ -61,7 +61,10 @@ def convert_constants_to_indices(rdp_constant_array: np.ndarray) -> np.ndarray:
     """
     # Find indices for all RDP constants <= 50
     sub50_mask = rdp_constant_array <= 50
-    sub50_indices = (((rdp_constant_array - 1) * sub50_mask) * 10_000).astype(int)
+    # np.maximum is to avoid negative indices when rdp_constant_array is < 1
+    sub50_indices = (
+        (np.maximum(0, rdp_constant_array - 1) * sub50_mask) * 10_000
+    ).astype(int)
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/adp/data_subject_ledger.py:67]: MAYBE: Could we do 
```py
sub50_mask = rdp_constant_array <= 50
sub50_indices = ((rdp_constant_array * sub50_mask) * 10_000).astype(int)
```
by removing the minus -1 
maybe that would have caused the negative indices?

File: packages/syft/src/syft/core/adp/data_subject_ledger.py [hunk 4]
```diff
@@ -61,7 +61,10 @@ def convert_constants_to_indices(rdp_constant_array: np.ndarray) -> np.ndarray:
...
 
     # Find indices for all RDP constants > 50
     gt50_mask = rdp_constant_array > 50
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/adp/data_subject_ledger.py:70]: MAYBE we could do:
```py
gt50_indices = ((rdp_constant_array - 50 + 500_000) * gt50_mask).astype(int)
```

File: packages/syft/src/syft/core/adp/data_subject_ledger.py (hunk 3 above)
Comment (IshanMi) [reply packages/syft/src/syft/core/adp/data_subject_ledger.py:67]: The minus 1 was definitely causing the negative indices, but it is necessary to get the correct indexing (see the image below)
image

If you look at the notebook here, it has more details: https://github.com/OpenMined/PySyft/blob/dev/notebooks/Experimental/Ishan/ExtendRDPCache.ipynb

Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_ledger.py:67]: Ahh , correct, 👍

File: packages/syft/src/syft/core/adp/data_subject_ledger.py [hunk 5]
```diff
@@ -61,7 +61,10 @@ def convert_constants_to_indices(rdp_constant_array: np.ndarray) -> np.ndarray:
     """
     # Find indices for all RDP constants <= 50
     sub50_mask = rdp_constant_array <= 50
-    sub50_indices = (((rdp_constant_array - 1) * sub50_mask) * 10_000).astype(int)
+    # np.maximum is to avoid negative indices when rdp_constant_array is < 1
+    sub50_indices = (
+        (np.maximum(0, rdp_constant_array - 1) * sub50_mask) * 10_000
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/adp/data_subject_ledger.py:66]: MAYBE: I think the -1 should be outside like
`np.maximum(0, (rdp_constant_array  * sub50_mask * 10_000)-1)`
Screenshot 2022-11-18 at 7 00 32 AM

Comment (IshanMi) [reply packages/syft/src/syft/core/adp/data_subject_ledger.py]: Great catch- fixed in the latest commit, will add a test to make sure the values are correct as well :+1:

File: packages/syft/tests/syft/core/adp/data_subject_ledger_test.py [hunk 6]
```diff
@@ -17,3 +18,16 @@ def test_data_subject_ledger_serde() -> None:
 
     assert de == dsl
     assert all(de._rdp_constants == rdp_constants)
+
+
+def test_cache_indexing_correctness() -> None:
```
Comment (rasswanth-s) [inline packages/syft/tests/syft/core/adp/data_subject_ledger_test.py:23]: ❤️

**Review by rasswanth-s — APPROVED** [review]
Well done @IshanMi , 
The DP system accuracy would be next level 🚀
